ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java

```diff
@@ -239,12 +248,243 @@ public void shouldBuildCorrectFormatsForV2NonWindowedTable() {
     ));
   }
 
+  @Test
+  public void shouldReplaceOldNonWindowedStreamSourceWithMatchingPseudoColumnVersion() {
+    // Given:
+    givenNonWindowedStream();
+    givenExistingQueryWithOldPseudoColumnVersion();
+    givenOnNewPseudoColumnVersion();
+
+    // When:
+    final SchemaKStream<?> result = SchemaKSourceFactory.buildSource(
+        buildContext,
+        dataSource,
+        contextStacker
+    );
+
+    // Then:
+    assertThat(((StreamSource) result.getSourceStep()).getPseudoColumnVersion(), equalTo(OLD_PSEUDOCOLUMN_VERSION));
+    assertValidSchema(result);
+  }
+
+  @Test
+  public void shouldCreateNonWindowedStreamSourceWithNewPseudoColumnVersionIfNoOldQuery() {
+    // Given:
+    givenNonWindowedStream();
+    givenOnNewPseudoColumnVersion();
+
+    // When:
+    final SchemaKStream<?> result = SchemaKSourceFactory.buildSource(
+        buildContext,
+        dataSource,
+        contextStacker
+    );
+
+    // Then:
+    assertThat(((StreamSource) result.getSourceStep()).getPseudoColumnVersion(), equalTo(NEW_PSEUDOCOLUMN_VERSION));
+    assertValidSchema(result);
+  }
+
+  @Test
+  public void shouldReplaceOldWindowedStreamSourceWithMatchingPseudoColumnVersion() {
+    // Given:
+    givenWindowedStream();
+    givenExistingQueryWithOldPseudoColumnVersion();
+    givenOnNewPseudoColumnVersion();
+
+    // When:
+    final SchemaKStream<?> result = SchemaKSourceFactory.buildSource(
+        buildContext,
+        dataSource,
+        contextStacker
+    );
+
+    // Then:
+    assertThat(((WindowedStreamSource) result.getSourceStep()).getPseudoColumnVersion(), equalTo(OLD_PSEUDOCOLUMN_VERSION));
+    assertValidSchema(result);
+  }
+
+  @Test
+  public void shouldCreateWindowedStreamSourceWithNewPseudoColumnVersionIfNoOldQuery() {
+    // Given:
+    givenWindowedStream();
+    givenOnNewPseudoColumnVersion();
+
+    // When:
+    final SchemaKStream<?> result = SchemaKSourceFactory.buildSource(
+        buildContext,
+        dataSource,
+        contextStacker
+    );
+
+    // Then:
+    assertThat(((WindowedStreamSource) result.getSourceStep()).getPseudoColumnVersion(), equalTo(NEW_PSEUDOCOLUMN_VERSION));
+    assertValidSchema(result);
+  }
+
+  @Test
+  public void shouldCreateNonWindowedTableSourceV2WithNewPseudoColumnVersionIfNoOldQuery() {
```

Comment:
Looks like there's no test checking that TableSourceV2 with an existing query uses the existing pseudocolumn version? Unless it throws an error to use an old version with TableSourceV2, I think it makes sense to add the test even if it's unrealistic at this time (you can leave a comment to point out the fact that it's unrealistic, if that makes it feel less weird).